aptos-move/move-examples/dao/nft_dao/sources/nft_dao.move

```diff
@@ -0,0 +1,660 @@
+    }
+
+    // transfer coin from the DAO account to the destination account
+    fun transfer_fund<CoinType>(res_acct: &signer, dst: address, amount: u64) {
+        transfer_coins<CoinType>(res_acct, dst, amount);
+    }
+
+    fun get_proposal_weights(
+        account: &signer,
+        token_names: &vector<String>,
+        property_versions: &vector<u64>,
+        dao: &DAO
+    ): u64 {
+        let gtoken = &dao.governance_token;
+        let i = 0;
+        let total_weight = 0;
+        let used_token_ids = vector::empty();
+
+        while (i < vector::length(token_names)) {
+            let token_name = *vector::borrow(token_names, i);
+            let property_version = *vector::borrow(property_versions, i);
+            let token_id = token::create_token_id_raw(gtoken.creator, gtoken.collection, token_name, property_version);
+            assert!(!vector::contains(&used_token_ids, &token_id), error::already_exists(ETOKEN_USED_FOR_CREATING_PROPOSAL));
+            vector::push_back(&mut used_token_ids, token_id);
+            assert!(token::balance_of(signer::address_of(account), token_id) == 1, error::permission_denied(ENOT_OWN_THE_VOTING_DAO_TOKEN));
+            total_weight = total_weight + 1;
+            i = i + 1;
+        };
+        total_weight
+    }
+
+    fun assert_function_valid(function_name: String, map: &PropertyMap){
+        if (function_name == string::utf8(b"transfer_fund")) {
+            assert!(property_map::length(map) == 2, error::invalid_argument(ENOT_SUPPROTED_FUNCTION));
+            property_map::read_address(map, &string::utf8(b"dst"));
+            property_map::read_u64(map, &string::utf8(b"amount"));
+        } else if (function_name == string::utf8(b"no_op")) {
+            assert!(property_map::length(map) == 0, error::invalid_argument(ENOT_SUPPROTED_FUNCTION));
+        } else {
+            abort error::invalid_argument(ENOT_SUPPROTED_FUNCTION)
+        }
+    }
+
+    #[test_only]
+    public fun create_one_token(
+        creator: &signer,
+        collection_name: String,
+        token_name: String,
+        amount: u64,
+        token_max: u64
+    ): TokenId {
+        create_token_script(
+            creator,
+            collection_name,
+            token_name,
+            string::utf8(b"Hello, Token"),
+            amount,
+            token_max,
+            string::utf8(b"https://aptos.dev"),
+            signer::address_of(creator),
+            100,
+            0,
+            vector<bool>[false, false, false, false, false],
+            vector::empty(),
+            vector::empty(),
+            vector::empty(),
+        );
+        token::create_token_id_raw(signer::address_of(creator), collection_name, token_name, 0)
+    }
+
+    #[test_only]
+    use aptos_token::token::create_token_script;
+    #[test_only]
+    use aptos_framework::coin::FakeMoney;
+    #[test_only]
+    use aptos_framework::coin;
+
+    #[test(aptos_framework = @0x1, creator = @0xdeaf, voter = @0xaf)]
```

Comment:
As a follow up, can you add more unit tests to verify all possible edge cases? Just a single test here (despite testing 2 flows e2e) seems insufficient.

Reply:
added to the linear tracking issue